libp2p/protocols/pubsub/pubsub.nim

```diff
@@ -581,37 +594,47 @@ method removeValidator*(
 
 method validate*(
     p: PubSub, message: Message
-): Future[ValidationResult] {.async, base.} =
+): Future[ValidationResult] {.async: (raises: [CancelledError]), base.} =
   var pending: seq[Future[ValidationResult]]
   trace "about to validate message"
   let topic = message.topic
   trace "looking for validators on topic",
     topic = topic, registered = toSeq(p.validators.keys)
   if topic in p.validators:
     trace "running validators for topic", topic = topic
-    for validator in p.validators[topic]:
-      pending.add(validator(topic, message))
+    try:
+      for validator in p.validators[topic]:
+        pending.add(validator(topic, message))
+    except KeyError:
+      raiseAssert "topic checked before"
```

Comment:
the alternate form that can be used is this:

```suggestion
    p.validators.withValue(topic, validators):
      for validator in validators[][topic]:
        pending.add(validator(topic, message))
```

This avoids having to do two lookups (one for `in` and one for the actual lookup - it also avoids the extra except KeyError. 

It only works when you have `var` access to the table however - here, it should work afair since `PubSub` is `ref`.